### PR TITLE
Add model instance update test

### PR DESCRIPTION
**Pull request by kthui**

### CODE REVIEW

**Comment by tanmayv25** [comment]
I believe you should also enhance L0_rate_limiter to test whether resource constraints are being updated properly upon scaling the instance counts up/down.

**Comment by kthui** [comment]
Rebased

**Comment by kthui** [comment]
Rebased.

**Comment by kthui** [comment]
Rebased.

File: qa/python_models/instance_init_del/util.py [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+def set_infer_delay(delay):
+    write_int_to_txt_file("instance_init_del_infer_delay.txt", delay)
+    return delay
+
+
+#
+# Functions for modifying the model
+#
+
+
```
Comment (tanmayv25) [inline qa/python_models/instance_init_del/util.py:128]: Can remove some of these extra blank lines.

File: qa/python_models/instance_init_del/util.py [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+def update_instance_group(instance_group_str):
```
Comment (tanmayv25) [inline qa/python_models/instance_init_del/util.py:129]: Can these functions go to general test utilities as they don't seem to be very specific for this model.

File: qa/python_models/instance_init_del/util.py [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+def reset_finalize_count():
+    return reset_count("instance_init_del_finalize_count.txt")
+
+
+#
+# Functions for communicating varies of delay (in seconds) to the model
+#
+
+
+def get_initialize_delay():
```
Comment (tanmayv25) [inline qa/python_models/instance_init_del/util.py:104]: How about `get_delay(kind)`?
where kind can be initialize/infer/finalize.

File: qa/python_models/instance_init_del/util.py [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+    delay = read_int_from_txt_file("instance_init_del_initialize_delay.txt")
+    return delay
+
+
+def set_initialize_delay(delay):
```
Comment (tanmayv25) [inline qa/python_models/instance_init_del/util.py:109]: Same as above, set_delay(kind)

File: qa/python_models/instance_init_del/util.py [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+    count = 0
+    write_int_to_txt_file(filename, count)
+    return count
+
+
+def get_initialize_count():
+    return get_count("instance_init_del_initialize_count.txt")
+
+
+def inc_initialize_count():
```
Comment (tanmayv25) [inline qa/python_models/instance_init_del/util.py:79]: inc_count(kind), where kind can be initialize/infer/finalize?

File: qa/L0_model_update/instance_update_test.py [hunk 6]
```diff
@@ -0,0 +1,310 @@
...
+        # Initialize client
+        self.__triton = grpcclient.InferenceServerClient("localhost:8001",
+                                                         verbose=True)
+
+    def __get_inputs(self, shape=(2,)):
+        inputs = [grpcclient.InferInput("INPUT0", shape, "FP32")]
+        inputs[0].set_data_from_numpy(np.ones(shape, dtype=np.float32))
+        return inputs
+
+    def test_add_rm_add(self):
```
Comment (tanmayv25) [inline qa/L0_model_update/instance_update_test.py:56]: Add a brief description about each of these test case.. What is the test_case testing specifically.

File: qa/L0_model_update/instance_update_test.py [hunk 7]
```diff
@@ -0,0 +1,310 @@
...
+        self.__triton.load_model("instance_init_del")
+        self.assertEqual(get_initialize_count(), 3)
+        self.assertEqual(get_finalize_count(), 0)
+        self.__triton.infer("instance_init_del", self.__get_inputs((8,)))
+        # Add 1 instance
+        update_instance_group("{\ncount: 4\nkind: KIND_CPU\n}")
+        self.__triton.load_model("instance_init_del")
+        self.assertEqual(get_initialize_count(), 4)
+        self.assertEqual(get_finalize_count(), 0)
+        self.__triton.infer("instance_init_del", self.__get_inputs((2,)))
```
Comment (tanmayv25) [inline qa/L0_model_update/instance_update_test.py:68]: Are we testing something specific here? We don't seem to be validating the results. I believe a better test would be to run 4 (number of active instance) async infer calls and check the results are available within specified time interval. (the inference occurred concurrently)

Comment (tanmayv25) [reply qa/L0_model_update/instance_update_test.py:68]: This can be done within a utility function that takes the number of instance counts.

File: qa/L0_model_update/instance_update_test.py [hunk 8]
```diff
@@ -0,0 +1,310 @@
...
+        self.__triton.unload_model("instance_init_del")
+        time.sleep(10)  # wait for unload to complete
+        self.assertEqual(get_initialize_count(), 5)
+        self.assertEqual(get_finalize_count(), 5)
+        with self.assertRaises(InferenceServerException):
+            self.__triton.infer("instance_init_del", self.__get_inputs((5,)))
+
+    def test_rm_add_rm(self):
+        # Load model
+        update_instance_group("{\ncount: 2\nkind: KIND_CPU\n}")
```
Comment (tanmayv25) [inline qa/L0_model_update/instance_update_test.py:91]: Another interesting case is to set the count to 0, send out infer request, there should be no responses as there are no instances available, then increase the instance count which would return some responses.

File: qa/L0_model_update/instance_update_test.py [hunk 9]
```diff
@@ -0,0 +1,310 @@
...
+        # Adding a new instance does not depend on existing instances, so the
+        # ongoing inference should not block the update.
+        self.assertGreaterEqual(infer_time, 10.0, "Invalid infer time")
+        self.assertLess(update_time, 5.0, "Update blocked by infer")
+        self.assertEqual(get_initialize_count(), 2)
+        self.assertEqual(get_finalize_count(), 0)
+        self.__triton.infer("instance_init_del", self.__get_inputs((8,)))
+        # Unload model
+        self.__triton.unload_model("instance_init_del")
+        time.sleep(10)  # wait for unload to complete
```
Comment (tanmayv25) [inline qa/L0_model_update/instance_update_test.py:263]: Instead of sleeping for 10 sec, can we poll the finalize_count every second for an entire period of 10s?
This will allow us to return earlier than 10sec.

File: qa/L0_model_update/instance_update_test.py [hunk 10]
```diff
@@ -0,0 +1,310 @@
...
+        # Waiting on new instance creation should not block inference on
+        # existing instances.
+        self.assertGreaterEqual(update_time, 10.0, "Invalid update time")
+        self.assertLess(infer_time, 5.0, "Infer blocked by update")
+        self.assertEqual(get_initialize_count(), 2)
+        self.assertEqual(get_finalize_count(), 0)
+        self.__triton.infer("instance_init_del", self.__get_inputs((8,)))
+        # Unload model
+        self.__triton.unload_model("instance_init_del")
+        time.sleep(10)  # wait for unload to complete
```
Comment (tanmayv25) [inline qa/L0_model_update/instance_update_test.py:302]: Same as above. These things can go to a utility function.

File: qa/L0_model_update/instance_update_test.py [hunk 11]
```diff
@@ -0,0 +1,310 @@
...
+        self.assertEqual(get_finalize_count(), 0)
+        self.__triton.infer("instance_init_del", self.__get_inputs((2,)))
+        # Infer while adding 1 instance
+        set_initialize_delay(10)
+        update_instance_group("{\ncount: 2\nkind: KIND_CPU\n}")
+        with concurrent.futures.ThreadPoolExecutor() as pool:
+            update_start_time = time.time()
+            update_thread = pool.submit(self.__triton.load_model,
+                                        "instance_init_del")
+            time.sleep(2)  # make sure update has started
```
Comment (tanmayv25) [inline qa/L0_model_update/instance_update_test.py:283]: Why? Didn't we want to test inference when update was in flight? I would assume a sequence of inference requests hitting the server when model instance update is in progress to be a better test.

**Review by tanmayv25 — COMMENTED** [review]
Add more testing on dynamic batcher and sequence batcher (both oldest and direct).
Add more testing on rate limiter resource limit updates to check model instance cross-prioritization occurs as expected.

File: qa/L0_model_update/instance_update_test.py (hunk 11 above)
Comment (kthui) [reply qa/L0_model_update/instance_update_test.py:283]: We test both: [test_update_while_inferencing](https://github.com/triton-inference-server/server/pull/5635/files#diff-40ca2af021f4288f3fc6ae6071a244bca6350a0e657ab2822b61650cac266931R230) and [test_infer_while_updating](https://github.com/triton-inference-server/server/pull/5635/files#diff-40ca2af021f4288f3fc6ae6071a244bca6350a0e657ab2822b61650cac266931R269)

File: qa/python_models/instance_init_del/util.py [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+
+
+def reset_count(filename):
+    count = 0
+    write_int_to_txt_file(filename, count)
+    return count
+
+
+def get_initialize_count():
+    return get_count("instance_init_del_initialize_count.txt")
```
[the remark this answers is not in the document]
Comment (kthui) [reply qa/python_models/instance_init_del/util.py]: Yes, the filenames are moved into variables. [Refactor python model_init_del](https://github.com/triton-inference-server/server/pull/5635/commits/10546c3ad4075a9791beddd64d6531a71627122d)

The name is actually formed by `[model_name]_[count_name_extension].txt`. In case another model in the same test would use the same count storage mechanism in the future, so their file names would not collide.

File: qa/python_models/instance_init_del/util.py [hunk 13]
```diff
@@ -0,0 +1,168 @@
...
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import os
+
+#
+# Helper functions for reading/writing state to disk
+#
+
+
+def read_int_from_txt_file(filename):
```
[the remark this answers is not in the document]
Comment (kthui) [reply qa/python_models/instance_init_del/util.py]: Yes. Updated: [Refactor python model_init_del](https://github.com/triton-inference-server/server/pull/5635/commits/10546c3ad4075a9791beddd64d6531a71627122d)

File: qa/python_models/instance_init_del/util.py
[the remark this answers is not in the document]
Comment (kthui) [reply qa/python_models/instance_init_del/util.py:1]: Thanks!

File: qa/python_models/instance_init_del/util.py (hunk 1 above)
Comment (kthui) [reply qa/python_models/instance_init_del/util.py]: They are from the formatter for Python, but I also think they are extra, so they are removed: [Refactor python model_init_del](https://github.com/triton-inference-server/server/pull/5635/commits/10546c3ad4075a9791beddd64d6531a71627122d)

File: qa/python_models/instance_init_del/util.py (hunk 2 above)
Comment (kthui) [reply qa/python_models/instance_init_del/util.py]: Yes, the task performed by the functions is not specific to this model, but the mechanism is simplified, which makes it specific to the `config.pbtxt` of the model. They can generate undefined result if used on other models. I think we can leave it here since it is only used here at this time, and make a generic implementation if another model needs it.

File: qa/python_models/instance_init_del/util.py (hunk 3 above)
Comment (kthui) [reply qa/python_models/instance_init_del/util.py]: Yes. Consolidated: [Util function consolidation](https://github.com/triton-inference-server/server/pull/5635/commits/cd601a939a97331ad1ad499cb2eb183e1eeea5bb)

File: qa/python_models/instance_init_del/util.py (hunk 4 above)
Comment (kthui) [reply qa/python_models/instance_init_del/util.py]: Consolidated: [Util function consolidation](https://github.com/triton-inference-server/server/pull/5635/commits/cd601a939a97331ad1ad499cb2eb183e1eeea5bb)

File: qa/python_models/instance_init_del/util.py (hunk 5 above)
Comment (kthui) [reply qa/python_models/instance_init_del/util.py]: Consolidated: [Util function consolidation](https://github.com/triton-inference-server/server/pull/5635/commits/cd601a939a97331ad1ad499cb2eb183e1eeea5bb)

File: qa/L0_model_update/instance_update_test.py (hunk 7 above)
Comment (kthui) [reply qa/L0_model_update/instance_update_test.py]: > I believe a better test would be to run 4 (number of active instance) async infer calls and check the results are available within specified time interval. (the inference occurred concurrently)

I do not think concurrent update will happen, because the `model_repository_manager` will serialize update (load/unload) to a model, which at best will only make the updates to be out-of-order.

> Are we testing something specific here?

Yes. This is to test if data structures can return to its original state by `original -> [add 1 item] original + 1 -> [remove 1 item] original -> [add 1 item] original + 1`. This is just to see if the implementation is well-formed (not to stress anything), because if something is missing, this simple case could break.

Comment (tanmayv25) [reply qa/L0_model_update/instance_update_test.py]: I meant the concurrent inference request and not concurrent updates.

File: qa/L0_model_update/instance_update_test.py (hunk 6 above)
Comment (kthui) [reply qa/L0_model_update/instance_update_test.py]: Yes. Updated: [Add short test description](https://github.com/triton-inference-server/server/pull/5635/commits/11c889fea337feeb59adb75246ebcbea6ff77d99)

File: qa/L0_model_update/instance_update_test.py (hunk 7 above)
Comment (kthui) [reply qa/L0_model_update/instance_update_test.py]: > utility function

Updated: [Refactor instance update test](https://github.com/triton-inference-server/server/pull/5635/commits/f012b0474ac4b1465e432e72f3c4f24cb5cb6b3c)

File: qa/L0_model_update/instance_update_test.py (hunk 9 above)
Comment (kthui) [reply qa/L0_model_update/instance_update_test.py]: Updated: [Refactor instance update test](https://github.com/triton-inference-server/server/pull/5635/commits/f012b0474ac4b1465e432e72f3c4f24cb5cb6b3c)

File: qa/L0_model_update/instance_update_test.py (hunk 10 above)
Comment (kthui) [reply qa/L0_model_update/instance_update_test.py]: Updated: [Refactor instance update test](https://github.com/triton-inference-server/server/pull/5635/commits/f012b0474ac4b1465e432e72f3c4f24cb5cb6b3c)

File: qa/L0_model_update/instance_update_test.py (hunk 8 above)
Comment (kthui) [reply qa/L0_model_update/instance_update_test.py]: Added a test verifying this behavior: [Add test for zero instance count](https://github.com/triton-inference-server/server/pull/5635/commits/124c594b01b986771b3b7c69df8936be3cc08d43)

File: qa/L0_model_update/instance_update_test.py [hunk 14]
```diff
@@ -0,0 +1,271 @@
...
+        self.__triton.load_model(self.__model_name)
+        self.assertEqual(get_count("initialize"), new_initialize_count)
+        if model_will_reload:
+            self.__poll_finalize_count(new_finalize_count)
+        self.assertEqual(get_count("finalize"), new_finalize_count)
+        self.__triton.infer(self.__model_name, self.__get_inputs(batching))
+
+    def __unload_model(self, batching=False):
+        prev_initialize_count = get_count("initialize")
+        prev_finalize_count = get_count("finalize")
```
[the remark this answers is not in the document]
Comment (kthui) [reply qa/L0_model_update/instance_update_test.py]: Fixed.

File: qa/L0_model_update/instance_update_test.py (hunk 7 above)
Comment (kthui) [reply qa/L0_model_update/instance_update_test.py]: > I meant the concurrent inference request and not concurrent updates.

Updated: [Add instance update with parallel infer](https://github.com/triton-inference-server/server/pull/5635/commits/27b9f9a0217b3a72df71a490541fd392e67d7b1b)

**Comment by kthui** [comment]
Added test for rate limiter resource update: [Add test for rate limiter resource update](https://github.com/triton-inference-server/server/pull/5635/commits/85f10344f820bba8729d5a57b363b7864978b0de)

**Comment by kthui** [comment]
Rebased.

**Review by tanmayv25 — APPROVED** [review]
